### DISCUSSION TITLE
Can I deploy a NextJS SSR App in Azure Functions?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [vercel/next.js#13289] [Help]
I'm needing a way to deploy a next app in serverless infra, there is a way to do yhis?

Thanks so much!

**Comment by wilsonneto-swe** [comment]
Great, I saw that but understood that this approach only runs on the Serverless Framework. But I will see with more attention and search more, thanks so much!!

**Reply by rafinskipg** [reply]
@wilsonneto-dev  did you find any more information?

**Reply by artem-alek** [reply]
@wilsonneto-dev @rafinskipg any more information? I am attempting to deploy a Next JS on azure but need a serverless function. We need incremental static gen and it seems that is not possible with a custom server per their documentations. I saw that azure released a Static Web App (preview) but it requires github and we use azure dev ops for all our code and release pipelines.

**Comment by artem-alek** [comment]
Any guides out there to create on azure that has all/close to all features of next js? Exporting or loosing ISG is not an option.

**Comment by steinarbragi** [comment]
I'm trying to get this working on Azure as well. I've deployed the site to Azure as a SWA, how do configure this to deploy API routes to Azure functions?

**Reply by csandman** [reply]
I am also in need of figuring out how to deploy the API routes

**Reply by lawrencejob** [reply]
I don't have the means to test this for myself right now, but I understand that this is a limitation of Azure Web Apps on _Windows_ (because of the iisnode module) and a Linux App Service might not need the server.js which breaks ASO and api routes.

**Comment by csandman** [comment]
I recently found a [post on Reddit](https://www.reddit.com/r/nextjs/comments/ix1adj/help_deploying_nextjs_to_azure_swa_with/) asking about this as well, and got into a conversation with someone about their setup.  They ended up open sourcing their solution [nextjs-azure-function](https://github.com/ShanonJackson/nextjs-azure-function) which expands upon the [serverless-next.js](https://github.com/serverless-nextjs/serverless-next.js) build process and outputs the api routes to a single Azure Function.  This is ideal for deploying to Azure Static Web Apps.

By the time he open sourced the package it was a little too late for me to switch to this process so I can't offer a lot of first hand support, however I thought this could probably help other people out!

**Reply by ShanonJackson** [reply]
I'm the author of that package and more than happy to help anyone trying to host NextJS applications across on prem / AWS and Azure; Serverless, Serverful, SSR, SSG etc.

We've been hosting applications outside for vercel for almost 3 years now and have honed some really good practices. I would recommend this for AWS https://registry.terraform.io/modules/dealmore/next-js/aws/latest which mimics vercel as closely as possible on AWS they have an entire blog post on why and how they did this.

On Azure we just straight up avoid SSG/SSR and use static-web-apps combined with next-export and some custom code and that handles all your default "jam stack" requirements (enterprise grade edge network, backend azure functions, security etc)

**Reply by superhp** [reply]
@ShanonJackson, that sounds promising. Do you use the "custom code" to export the SSR part into Azure Functions while other parts of the project are exported with next-export/SSG into Azure Static Web App? As I see in your project repository, the SSG part is still not supported.

**Comment by aaronpowell** [comment]
I know it's been a while since this issue was opened but today, we announced support in [Azure Static Web Apps for Next.js SSR (and other "server" features)](https://techcommunity.microsoft.com/t5/apps-on-azure-blog/extending-next-js-support-in-azure-static-web-apps/ba-p/3627975).

While this isn't directly Azure Functions, I don't think that Azure Functions isn't really the right platform for this sort of hosting as it's designed for serverless event-based responses, where Static Web Apps is for hosting frontend-centric web applications.

**Reply by lukaesch** [reply]
@aaronpowell When you deploy a hybrid Next.js app on Azure Static Web Apps using the "Enterprise-grade edge"-feature are you rendering the server-side generated pages on the edge?

**Reply by aaronpowell** [reply]
I'm not sure, let me check with the product team and get back to you.

**Reply by morganfeeney** [reply]
Hey @aaronpowell did you manage to get this info?

**Reply by aaronpowell** [reply]
Sorry, I did but forgot to update.

Some aspects of Next.js will be cached with this feature, SSG and ISR files, but SSR content won't be.

**Reply by colinscz** [reply]
@aaronpowell the preview mode for nextjs SSR deployment on Static Web Apps looks promising, is there any timeline on when this feature for SSR will be labeled production ready?

**Reply by thomasgauvin** [reply]
@colinscz we're planning on shipping NextJS support on Azure Static Web Apps as GA by end of summer

**Reply by colinscz** [reply]
@thomasgauvin thank you for the rough timeline - I am looking forward to the release 👍🏽

**Reply by harkins-it** [reply]
@thomasgauvin we are past summer. Is there any update on this being in GA?

**Comment by zpg6** [comment]
I just added native serverless Next.js to Azure Functions. This project brings the same Vercel-grade developer experience to Azure: one command deploys your Next.js app with all the infrastructure you need. Supports ISR, streaming SSR, and on-demand revalidation.

Would love any feedback you can provide and please feel free to contribute!

https://github.com/zpg6/opennextjs-azure
